src/utils/daos_hdlr.c

```diff
@@ -1096,6 +1096,136 @@ cont_get_acl_hdlr(struct cmd_args_s *ap)
 	return rc;
 }
 
+/*
+ * Returns a substring of the line with leading and trailing whitespace trimmed.
+ * Doesn't allocate any new memory - trimmed string is just a pointer.
+ */
+static char *
+trim_acl_file_line(char *line)
+{
+	char *end;
+
+	while (isspace(*line))
+		line++;
+	if (line[0] == '\0')
+		return line;
+
+	end = line + strnlen(line, DAOS_ACL_MAX_ACE_STR_LEN) - 1;
+	while (isspace(*end))
+		end--;
+	end[1] = '\0';
+
+	return line;
+}
+
+static int
+parse_acl_file(const char *path, struct daos_acl **acl)
+{
+	int		rc = 0;
+	FILE		*instream;
+	char		*line = NULL;
+	size_t		line_len = 0;
+	char		*trimmed;
+	struct daos_ace	*ace;
+	struct daos_acl	*tmp_acl;
+
+	instream = fopen(path, "r");
+	if (instream == NULL) {
+		fprintf(stderr, "Unable to read ACL input file '%s': %s\n",
+			path, strerror(errno));
+		return daos_errno2der(errno);
+	}
+
+	tmp_acl = daos_acl_create(NULL, 0);
+	if (tmp_acl == NULL) {
+		fprintf(stderr, "Unable to allocate memory for ACL\n");
+		D_GOTO(out, rc = -DER_NOMEM);
+	}
+
+	while (getline(&line, &line_len, instream) != -1) {
+		trimmed = trim_acl_file_line(line);
+
+		/* ignore blank lines and comments */
+		if (trimmed[0] == '\0' || trimmed[0] == '#') {
+			D_FREE(line);
+			continue;
+		}
+
+		rc = daos_ace_from_str(trimmed, &ace);
+		if (rc != 0) {
+			fprintf(stderr,
+				"Error parsing ACE '%s' from file: %s (%d)\n",
+				trimmed, d_errstr(rc), rc);
+			D_GOTO(parse_err, rc);
+		}
+
+		rc = daos_acl_add_ace(&tmp_acl, ace);
+		daos_ace_free(ace);
+		if (rc != 0) {
+			fprintf(stderr, "Error parsing ACL file: %s (%d)\n",
+				d_errstr(rc), rc);
+			D_GOTO(parse_err, rc);
+		}
+
+		D_FREE(line);
+	}
+
+	if (daos_acl_validate(tmp_acl) != 0) {
+		fprintf(stderr, "Content of ACL file is invalid\n");
+		daos_acl_free(tmp_acl);
+		D_GOTO(out, rc = -DER_INVAL);
+	}
+
+	*acl = tmp_acl;
+
+out:
+	fclose(instream);
+	return rc;
+
+parse_err:
```

Comment:
can this parse_err block (except the fclose and return rc lines) be moved up in front of the out: label?

Reply:
We could change the order of the blocks, but that would mean we'd need to use a goto in the success case, which struck me as a little weird. We definitely don't want to free tmp_acl in the success case, since we're returning it.

Reply:
Oh yes my mistake on that.

Reply:
Went ahead and moved it. Please have a look at the new version and see what you think.